fairseq/models/t5/t5_modules.py

```diff
@@ -0,0 +1,516 @@
+"""
+Pytorch modules for T5 model.
+
+The code based on:
+ * https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_t5.py
+
+The original code is licensed under Apache 2.0.
+
+"""
+
+import itertools
+import math
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+
+class T5LayerNorm(nn.Module):
+    def __init__(self, hidden_size, eps=1e-6):
+        """Construct a layernorm module in the T5 style (no bias and no substraction of mean)
+
+        Args:
+        hidden_size: size of a hidden state
+        eps: epsilon to use to variance normalization
+
+        """
+        super().__init__()
+        self.weight = nn.Parameter(torch.ones(hidden_size))
+        self.variance_epsilon = eps
+
+    def forward(self, x):
+        variance = x.pow(2).mean(-1, keepdim=True)
+        x = x / torch.sqrt(variance + self.variance_epsilon)
+        return self.weight * x
+
+
+class T5DenseReluDense(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.wi = nn.Linear(config.d_model, config.d_ff, bias=False)
+        self.wo = nn.Linear(config.d_ff, config.d_model, bias=False)
+        self.dropout = nn.Dropout(config.dropout_rate)
+
+    def forward(self, hidden_states):
+        h = self.wi(hidden_states)
+        h = F.relu(h)
+        h = self.dropout(h)
+        h = self.wo(h)
+        return h
+
+
+class T5LayerFF(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.DenseReluDense = T5DenseReluDense(config)
+        self.layer_norm = T5LayerNorm(config.d_model, eps=config.layer_norm_epsilon)
+        self.dropout = nn.Dropout(config.dropout_rate)
+
+    def forward(self, hidden_states):
+        norm_x = self.layer_norm(hidden_states)
+        y = self.DenseReluDense(norm_x)
+        layer_output = hidden_states + self.dropout(y)
+        return layer_output
+
+
+class T5Attention(nn.Module):
+    NEW_ID = itertools.count()
+
+    def __init__(self, config, has_relative_attention_bias=False, is_cross_attention=False):
```

Comment:
Hello, this file is based on https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_t5.py (most like of commit 
https://github.com/huggingface/transformers/commit/bbf26c4e619cf42106163e1e2cd5ff98b936ff93#diff-7370db3a19209bf984cc40925aaf2b71 as I binary searched).

While I found there a few modifications, by intention. This is_cross_attention is one of them. If is_cross_attention is True (For  EncDecAttention), relative_attention_bias will use nn.Functional.embedding(rp_bucket, torch.zeros(...)) 
 instead of nn.Embedding(). The problem I observed here is: in the former one, the embedding dict is not trainable, they are all zero. I checked the  https://applica-public.s3-eu-west-1.amazonaws.com/fairseq-t5/t5-small.pt, which did not contain the relative_attention_bias for encdecattentions. But I do find the weight on latest huggingface https://s3.amazonaws.com/models.huggingface.co/bert/t5-small-pytorch_model.bin. 



1. 

)

Reply:
That's a great question. The huggingface checkpoint is created in the following way: initialize the whole model randomly and then overwrite weights from Google checkpoint. Thus, the huggingface checkopoint contains EncDec relative embeddings (that are not in Google one). I can provide a simple script I used to convert from Tensorflow to Fairseq.

Lastly, I started working on the unsupervised task (I should have it tested until the end of this month). In addition, I discovered that I did not implement keeping decoder state during decoding, so it is a bit slow if the generated text is long.

Reply:
Thanks @tomekd. I assume the script converting from TF ckpt to Fairseq also contains the necessary optimizer states besides the weights themselves, right ? If you can share with that, that would be helpful. Thanks a lot!

Reply:
Sure thing! I will do it later today.

Reply:
Here is the script I used to convert: [script](https://gist.github.com/tomekd/899adcfae511fb3a2175488e812dbd32) Let me in case of any problem. Once I converted the models, I did not use it anymore later.

Reply:
Thanks you so much!